aloha-core/src/main/scala/com/eharmony/aloha/dataset/vw/multilabel/VwDownsampledMultilabelRowCreator.scala

```diff
@@ -0,0 +1,378 @@
+    * '''This should not be used directly.''' ''It is exposed for testing.''
+    *
+    * @param features the common features produced
+    * @param indices indices of all labels in the training set.
+    * @param positiveLabelIndices indices of labels in the training set that positive are
+    *                             positive for this training example
+    * @param defaultNs list of feature indices in the default VW namespace.
+    * @param namespaces a mapping from VW namespace name to feature indices in that namespace.
+    * @param classNs the namespace name for class information.
+    * @param dummyClassNs the namespace name for dummy class information.  2 dummy classes are
+    *                     added to make the predicted probabilities work.
+    * @param negativeDummyStr line in VW input associated with the negative dummy class.
+    * @param positiveDummyStr line in VW input associated with the positive dummy class.
+    * @param seed a seed for randomness.  The second part of the output of this function is a
+    *             new seed that should be used on the next call to this function.
+    * @param numNegLabelsTarget the desired number of negative labels.  If it is determined
+    *                           that there are less negative labels than desired, no negative
+    *                           label downsampling will occur; otherwise, the negative labels
+    *                           will be downsampled to this target value.
+    * @return an array representing the VW input that can either be passed directly to the
+    *         VW JNI library, or `mkString("", "\n", "\n")` can be called to pass to the VW
+    *         CLI.  The second part of the return value is a new seed to use as the `seed`
+    *         parameter in the next call to this function.
+    */
+  private[multilabel] def sampledTrainingInput(
+      features: IndexedSeq[Sparse],
+      indices: sci.IndexedSeq[Int],
+      positiveLabelIndices: Int => Boolean,
+      defaultNs: List[Int],
+      namespaces: List[(String, List[Int])],
+      classNs: Char,
+      dummyClassNs: Char,
+      negativeDummyStr: String,
+      positiveDummyStr: String,
+      seed: Long,
+      numNegLabelsTarget: Int
+  ): (Array[String], Long) = {
+
+    // Partition into positive and negative indices.
+    val (posLabelInd, negLabelInd) = indices.partition(positiveLabelIndices)
+    val negLabels = negLabelInd.size
+
+    val numDownsampledLabels = math.min(numNegLabelsTarget, negLabels)
+
+    // Sample indices in negLabelInd to use in the output.
+    val (downsampledNegLabelInd, newSeed) =
+      sampleCombination(negLabels, numDownsampledLabels, seed)
+
+    val p = posLabelInd.size
+    val n = downsampledNegLabelInd.length
+
+    // This code is written with knowledge of the constant's value.
+    // TODO: Write calling code to abstract over costs so knowledge of the costs isn't necessary.
+    val negWt =
+      if (numDownsampledLabels == negLabels) {
```

Comment:
Should this be `==` or `<=`?

Reply:
`==`.  Since `numDownsampledLabels == math.min(numNegLabelsTarget, negLabels)`, it is certain that `numDownsampledLabels <= negLabels`.  We only don't downsample when `numDownsampledLabels == negLabels`.